### PR TITLE
Fix validation by moving it into the resource builder.

### LINKED ISSUE #7598: kubectl --validate flag doesn't complain about unknown fields
My understanding of kubectl's --validate flag is that it's supposed to error out if it can't parse its input as a valid API object. Maybe my understanding is wrong, but if it isn't, then the --validate flag is currently broken.

This came up when a user's service was coming up without a load balancer when they used the following yaml:
```
kind: Service
apiVersion: v1beta3
metadata:
  name: locust-master
  labels:
    name: locust
    role: master
spec:
  ports:
    - port: 8089
      targetPort: 8089
      protocol: TCP
      name: locust-master-web
  selector:
    name: locust
    role: master
createExternalLoadBalancer: true
```

The createExternalLoadBalancer parameter is incorrectly at the top level, when it's supposed to be inside the spec. Running `kubectl create -f locust-master-service.yaml --validate` happily succeeds, with the server never seeing the createExternalLoadBalancer parameter.

You can make this even more extreme by throwing in arbitrarily named fields that don't exist anywhere in the API:
```
kind: Service
apiVersion: v1beta3
metadata:
  xxxxxxxxx: yyyyyyyy
  name: locust-master
  labels:
    name: locust
    role: master
spec:
  aaaaaaaa: bbbbbbbbb
  ports:
    - port: 8089
      targetPort: 8089
      protocol: TCP
      name: locust-master-web
  selector:
    name: locust
    role: master
```

Running `kubectl create -f locust-master-service.yaml --validate` with this input file also happily succeeds.

@brendandburns @jlowdermilk

### PR DESCRIPTION
**Pull request by brendandburns**
Also always print an error for unknown field.

Closes: https://github.com/GoogleCloudPlatform/kubernetes/issues/7598

@a-robinson @ironcladlou @deads2k

### CODE REVIEW

File: pkg/kubectl/resource/visitor.go [hunk 1]
```diff
@@ -180,6 +181,20 @@ func (l EagerVisitorList) Visit(fn VisitorFunc) error {
...
+func ValidateSchema(data []byte, schema validation.Schema) error {
+	if schema == nil {
+		return nil
+	}
+	data, err := yaml.ToJSON(data)
+	if err != nil {
+		return fmt.Errorf("error converting to YAML: %v", err)
+	}
+	if err := schema.ValidateBytes(data); err != nil {
+		return fmt.Errorf("Error validating data: %v", err)
```
Comment (ironcladlou) [inline pkg/kubectl/resource/visitor.go]: Trivial: lowercase "error" for consistency.

File: pkg/kubectl/cmd/create.go [hunk 2]
```diff
@@ -78,7 +78,21 @@ func RunCreate(f *cmdutil.Factory, out io.Writer, filenames util.StringList) err
...
+			return err
+		}
+		data, err = yaml.ToJSON(data)
+		if err != nil {
+			return err
+		}
+		if err := schema.ValidateBytes(data); err != nil {
+			return err
+		}
+	}*/
```
Comment (ironcladlou) [inline pkg/kubectl/cmd/create.go]: Minor: dead code.

**Comment by ironcladlou** [comment]
I'm not familiar with the stuff being changed here, so my comments are pretty limited.

File: pkg/kubectl/cmd/create.go (hunk 2 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/create.go]: done.

File: pkg/kubectl/resource/visitor.go (hunk 1 above)
Comment (brendandburns) [inline pkg/kubectl/resource/visitor.go]: done.

**Comment by brendandburns** [comment]
Comments addressed.  @smarterclayton is there someone else from Red Hat who I should add to look at this stuff?

**Comment by a-robinson** [comment]
Thanks for doing this, Brendan! Assigning Jeff since he's much more familiar with the code here.

**Comment by smarterclayton** [comment]
Me, probably.

File: pkg/kubectl/resource/visitor.go [hunk 3]
```diff
@@ -386,8 +414,8 @@ type StreamVisitor struct {
 // empty stream is treated as an error for now.
 // TODO: convert ignoreErrors into a func(data, error, count) bool that consumers can use to decide
 // what to do with ignored errors.
-func NewStreamVisitor(r io.Reader, mapper *Mapper, source string, ignoreErrors bool) Visitor {
-	return &StreamVisitor{r, mapper, source, ignoreErrors}
+func NewStreamVisitor(r io.Reader, mapper *Mapper, schema validation.Schema, source string, ignoreErrors bool) Visitor {
```
Comment (smarterclayton) [inline pkg/kubectl/resource/visitor.go]: It's unfortunate having to decorate all the visitors like this.  There is no other way today unless we instead made this a generic interface (that allows you to act on the undecoded data prior to operating on it).

File: pkg/kubectl/resource/visitor.go [hunk 4]
```diff
@@ -386,8 +414,8 @@ type StreamVisitor struct {
...
+	return &StreamVisitor{r, mapper, source, ignoreErrors, schema}
```
Comment (smarterclayton) [inline pkg/kubectl/resource/visitor.go]: You need to add this to the streamVisitor Visit method

File: pkg/kubectl/resource/visitor.go (hunk 3 above)
Comment (brendandburns) [inline pkg/kubectl/resource/visitor.go]: acknowledged.

File: pkg/kubectl/resource/visitor.go (hunk 4 above)
Comment (brendandburns) [inline pkg/kubectl/resource/visitor.go]: done.

**Comment by brendandburns** [comment]
Comments addressed.  Please take another look.

Thanks!
--brendan

File: pkg/kubectl/resource/visitor.go [hunk 5]
```diff
@@ -285,6 +311,7 @@ func (v *URLVisitor) Visit(fn VisitorFunc) error {
 	if err != nil {
 		return fmt.Errorf("unable to read URL %q: %v\n", v.URL, err)
 	}
+	ValidateSchema(data, v.Schema)
```
Comment (jlowdermilk) [inline pkg/kubectl/resource/visitor.go]: Did you mean to throw away the return value here?

Comment (brendandburns) [inline pkg/kubectl/resource/visitor.go]: done.

**Comment by brendandburns** [comment]
Sorry for the lack of tests, they're not pre-existing for these classes, and doing all of the scheme/mapper/mapping classes is not simple.

--brendan

**Comment by smarterclayton** [comment]
LGTM

**Comment by jlowdermilk** [comment]
lgtm

**Comment by nikhiljindal** [comment]
Travis is green. Merging.

**Pull request merged**